walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,155 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
+  /**
+   * Given a Collection of all known events and a meeting request, returns the TimeRanges where
+   * the meeting can occur.
+   * 
+   * Algorithm:
+   * 1. Going through all the events, add events that involve meeting attendees (both optional and 
+   * mandatory) to an ArrayList, eventList.
+   * 2. Sort eventList according to event start times
+   * 3. Merge adjacent events if they overlap
+   * 4. Going through the merged events list, add the gaps between events that are longer or 
+   * equal to meeting duration to output
+   * 5. If there are no meeting times, remove optional attendees, and repeat steps 1-4.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    ArrayList<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+    // EDGE CASE: No meetingTimes available for meetings longer than 1 day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return meetingTimes;
+    }
+    
+    /* FILTER ALL EVENTS BY ATTENDEES */
+    // Initially consider both mandatory and optional attendees
+    HashSet<String> meetingAttendees = new HashSet<String>(request.getAttendees());
+    meetingAttendees.addAll(request.getOptionalAttendees());
+    ArrayList<Event> eventList = filterEventsByAttendees(events, meetingAttendees);
+
+    /* SORT EVENTS BY START TIME */ 
+    Comparator<Event> compareByStartTime = 
+        (Event event1, Event event2) -> Integer.compare(event1.getWhen().start(), event2.getWhen().start());
+    eventList.sort(compareByStartTime);
+
+    /* MERGE ANY OVERLAPPING EVENTS */ 
+    ArrayList<TimeRange> mergedEventTimes = mergeOverlappingEvents(eventList);
+    
+    /* RETURN ALL THE BLOCKS OF AVAILABILITY */
+    meetingTimes = findAvailability(mergedEventTimes, request.getDuration());
+
+    // If at this point there are no possible meeting times, try removing optional guests
+    // and repeat the process.
+    if (meetingTimes.isEmpty()) {
+      meetingAttendees.removeAll(request.getOptionalAttendees());
+      eventList = filterEventsByAttendees(events, meetingAttendees);
+      eventList.sort(compareByStartTime);
+      mergedEventTimes = mergeOverlappingEvents(eventList);
+      meetingTimes = findAvailability(mergedEventTimes, request.getDuration());
+    }
+    return meetingTimes;
+  }
+
+  /**
+   * Returns the subset of events in which the attendees provided are attending.
+   * @param events the collection of Event objects to be filtered
+   * @param attendees the attendees to filter the Event objects by
```

Comment:
missing an `@return` here

Reply:
Done! 
I was thinking that the method description already covered what the method will return, which is why I thought maybe an @return statement was not necessary, but I guess it is the best practice for Javadoc.